File: Mobzy-API/src/main/java/com/offz/spigot/mobzy/Pathfinders/PathfinderGoalLookAtPlayerPitchLock.java
package com.offz.spigot.mobzy.Pathfinders;

import net.minecraft.server.v1_13_R2.*;

public class PathfinderGoalLookAtPlayerPitchLock extends PathfinderGoal {
    protected EntityInsentient entity; //know for sure
    protected Entity lookAt; //know for sure
    protected float c;
    private int e;
    private final float startChance; //know for sure
    protected Class<? extends Entity> d;

    public PathfinderGoalLookAtPlayerPitchLock(EntityInsentient var1, Class<? extends Entity> var2, float var3) {
        this(var1, var2, var3, 0.02F);
    }

    public PathfinderGoalLookAtPlayerPitchLock(EntityInsentient var1, Class<? extends Entity> var2, float var3, float var4) {
        this.entity = var1;
        this.d = var2;
        this.c = var3;
        this.startChance = var4;
        this.a(2);
    }

    public boolean a() {
        if (this.entity.getRandom().nextFloat() >= this.startChance) {
            return false;
        } else {
            if (this.entity.getGoalTarget() != null) {
                this.lookAt = this.entity.getGoalTarget();
            }

            if (this.d == EntityHuman.class) {
                this.lookAt = this.entity.world.a(this.entity.locX, this.entity.locY, this.entity.locZ, (double)this.c, IEntitySelector.f.and(IEntitySelector.b(this.entity)));
            } else {
                this.lookAt = this.entity.world.a(this.d, this.entity.getBoundingBox().grow((double)this.c, 3.0D, (double)this.c), this.entity);
            }

            return this.lookAt != null;
        }
    }

    public boolean b() {
        if (!this.lookAt.isAlive()) {
            return false;
        } else if (this.entity.h(this.lookAt) > (double)(this.c * this.c)) {
            return false;
        } else {
            return this.e > 0;
        }
    }

    public void c() {
        this.e = 40 + this.entity.getRandom().nextInt(40);
    }

    public void d() {
        this.lookAt = null;
    }

    public void e() {
//        this.entity.getControllerLook().a(this.lookAt.locX, this.lookAt.locY + (double)this.lookAt.getHeadHeight(), this.lookAt.locZ, (float)this.entity.L(), (float)this.entity.K());
        this.entity.getControllerLook().a(this.lookAt.locX, this.entity.locY + (double)this.entity.getHeadHeight(), this.lookAt.locZ, (float)this.entity.L(), (float)this.entity.K());
        --this.e;
    }
}

File: Mobzy-API/src/main/java/com/offz/spigot/mobzy/pathfinders/PathfinderGoalLookAtPlayerPitchLock.kt
package com.offz.spigot.mobzy.pathfinders

import com.offz.spigot.mobzy.mobs.CustomMob
import com.offz.spigot.mobzy.toNMS
import net.minecraft.server.v1_15_R1.EntityInsentient
import net.minecraft.server.v1_15_R1.EntityTypes
import org.bukkit.entity.LivingEntity
import kotlin.random.Random

class PathfinderGoalLookAtPlayerPitchLock(
        private var mob: CustomMob,
        private val targetType: EntityTypes<*>,
        private val radius: Double,
        private val startChance: Float = 0.02f) : com.offz.spigot.mobzy.pathfinders.PathfinderGoal() {
    var entity: LivingEntity = mob.entity.bukkitEntity as LivingEntity
    var nmsEntity: EntityInsentient = mob.entity as EntityInsentient
    var lookAt: org.bukkit.entity.Entity? = null
    private var length = 0

    override fun shouldExecute(): Boolean {
        if (Random.nextFloat() >= startChance) return false
        if (nmsEntity.goalTarget != null) {
            lookAt = nmsEntity.goalTarget!!.living
            return true
        }

        lookAt = entity.getNearbyEntities(radius, radius, radius)
                .ifEmpty { return false }
                .filter { other -> other!!.toNMS().entityType === targetType }
                .minBy { mob.distanceToEntity(it) }
//        nmsEntity.goalTarget = lookAt!!.toNMS() as EntityLiving
        return true
    }

    override fun shouldTerminate(): Boolean {
        if(lookAt == null) return true
        return if (lookAt!!.isDead ||
                mob.distanceToEntity(lookAt!!) > (radius * radius)) false
        else length > 0
    }

    override fun init() {
        length = Random.nextInt(40, 80)
    }

    override fun reset() {
        lookAt = null
    }

    override fun execute() {
        if(lookAt == null) return
        mob.lookAtPitchLock(lookAt!!)
        --length
    }
}